spec/support/lti_controller_examples.rb

```diff
@@ -0,0 +1,136 @@
+shared_examples 'lti deployment controller' do
+  let(:instructor) { create :instructor }
+  let!(:client_id) { 'LMS defined ID' }
+  let(:target_link_uri) { 'https://example.com/authorize_redirect' }
+  let(:host) { 'https://canvas.instructure.com' }
+  let(:state) { 'state_param' }
+  describe '#launch' do
+    context 'when launching with invalid parameters' do
+      let(:lti_message_hint) { 'opaque string' }
+      let(:login_hint) { 'another opque string' }
+      let(:auth_url) { 'http://canvas.instructure.com:433/api/lti/authorize_redirect' }
+      it 'responds with unprocessable_entity if no parameters are passed' do
+        post :launch, params: {}
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      it 'responds with unprocessable_entity if lti_message_hint is not passed' do
+        post :launch, params: { client_id: client_id, target_link_uri: target_link_uri, login_hint: login_hint }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      it 'responds with unprocessable_entity if client_id is not passed' do
+        post :launch,
+             params: { lti_message_hint: lti_message_hint, target_link_uri: target_link_uri, login_hint: login_hint }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      it 'responds with unprocessable_entity if target_link_uri is not passed' do
+        post :launch, params: { lti_message_hint: lti_message_hint, client_id: client_id, login_hint: login_hint }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      it 'responds with unprocessable_entity if login_hint is not passed' do
+        post :launch,
+             params: { lti_message_hint: lti_message_hint, client_id: client_id, target_link_uri: target_link_uri }
+        is_expected.to respond_with(:unprocessable_entity)
+      end
+      context 'when all required params exist' do
+        before :each do
+          stub_request(:post, 'http://canvas.instructure.com:443/api/lti/authorize_redirect')
```

Comment:
if this is a shared example for all lti controller subclasses, do not use canvas specific examples